### ISSUE TITLE
`{usage}` expansion in custom usage no longer works

### ISSUE THREAD

**Issue by scottlamb** [pacak/bpaf#222]
Since 0.9.0's introduction of `bpaf::doc::Doc`, I don't see how to do the `{usage}` expansion that used to be possible. There are a couple examples in the repo, but they're broken.

(btw, looks like the `v0.9.0` tag doesn't appear to have been pushed yet)

```
$ git describe --dirty
v0.8.1-162-gcd5920f
$ rg '\{usage\}'
bpaf_derive/src/top_tests.rs
95:        #[bpaf(options, usage("App: {usage}"))]
107:                .usage("App: {usage}")

examples/customize_help.rs
14:        .usage("You can call it with following flags: {usage}")

Changelog.md
228:- templated usage string: can use "{usage}" in custom overrides
$ cargo run --example customize_help -- --help
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/examples/customize_help --help`
I am a program and I do things

You can call it with following flags: {usage}

Sometimes they even work.

Available options:
    -d          Release the dragon
    -h, --help  Prints help information

Beware `-d`, dragons be here
```

**Comment by pacak** [comment]
I knew I forgot something :) 

Main motivation for `{usage}` expansion was to include application name in the command line, now this is done automatically so most people won't need it. For those who still want it - there's `with_usage` which gives you a document for original usage and lets you compose the new usage string using that. Please let me know if something is still missing.

https://docs.rs/bpaf/latest/bpaf/struct.OptionParser.html#method.with_usage

I'll look into updating examples.

**Comment by pacak** [comment]
Tags should be up to date, at least pushing them one more time does nothing...

```console
% git push origin v0.9.0 bpaf_derive-v0.5.0
Everything up-to-date
```

https://github.com/pacak/bpaf/tags

**Comment by scottlamb** [comment]
> I knew I forgot something :)

I know that feeling well...

> Main motivation for `{usage}` expansion was to include application name in the command line, now this is done automatically so most people won't need it. For those who still want it - there's with_usage which gives you a document for original usage and lets you compose the new usage string using that. Please let me know if something is still missing.

Oh, I missed `with_usage`, will try, thanks.

I'm not seeing the application name in the command line by default, though, e.g. if I omit my custom usage (https://github.com/scottlamb/moonfire-nvr/commit/4cbaf178b0ff8f2064650404d0b1a32182fe3698) I just see this:

```
Moonfire NVR: security camera network video recorder.

Usage: COMMAND ...
```

**Comment by scottlamb** [comment]
> Tags should be up to date, at least pushing them one more time does nothing...

Oh, I think it's that unlike earlier tags, it's not an "annotated" tag, so `git describe` doesn't use it by default.

**Comment by pacak** [comment]
It will not show up in the tests because tests don't know the app name (unless you specify it yourself with `set_name`, but it will show up when you run it. That's what I see when I run the link you posted.

![1](https://github.com/pacak/bpaf/assets/66211/446c6a88-5252-4435-aaa9-3f46c02259f6)

https://docs.rs/bpaf/latest/bpaf/struct.Args.html#method.set_name

**Comment by pacak** [comment]
There's a few things you can simplify a bit with new features:

```rust
fn parse_db_dir() -> impl Parser<std::path::PathBuf> {
    bpaf::long("db-dir")
        .help(format!(
            "Directory holding the SQLite3 index database.\nDefault: `{}`",
            DEFAULT_DB_DIR
        ))
        .argument::<std::path::PathBuf>("PATH")
        .fallback_with(|| Ok::<_, std::convert::Infallible>(DEFAULT_DB_DIR.into()))
}
```

```rust
fn parse_db_dir() -> impl Parser<std::path::PathBuf> {
    bpaf::long("db-dir")
        .help("Directory holding the SQLite3 index database.")
        .argument::<std::path::PathBuf>("PATH")
        .fallback(DEFAULT_DB_DIR.into())
        .debug_fallback()
}
```

Hmm... I wonder it it makes sense to add `debug_fallback` to `fallback_with`...

**Comment by scottlamb** [comment]
> Hmm... I wonder it it makes sense to add `debug_fallback` to `fallback_with`...

Makes sense to me, but honestly I was probably prematurely optimizing by using `fallback_with` anyway. It's not as if one unnecessary `PathBuf` allocation per run (in the relatively rare case where someone actually overrides this default) is going to be noticeable.

**Comment by pacak** [comment]
Well, with regular `fallback` there's always a fallback value I can display, `fallback_with` is allowed to fail...
